### PR TITLE
Feat/sw reload only restart chaned

### PR DESCRIPTION
**Pull request by xbsura**
after https://github.com/VictoriaMetrics/VictoriaMetrics/pull/459

sw now only restart if changed, to aviod scrape interval reset after restarting

### CODE REVIEW

File: lib/promscrape/scraper.go [hunk 1]
```diff
@@ -64,7 +67,7 @@ func runScraper(configFile string, pushData func(wr *prompbmarshal.WriteRequest)
 	signal.Notify(sighupCh, syscall.SIGHUP)
 
 	logger.Infof("reading Prometheus configs from %q", configFile)
-	cfg, data, err := loadConfig(configFile)
+	cfg, _, err := loadConfig(configFile)
```
Comment (xbsura) [inline lib/promscrape/scraper.go:70]: as all sws can be restarted smarter, check whether configFile change is not necessary now

File: lib/promscrape/config_test.go [hunk 2]
```diff
@@ -97,7 +97,7 @@ scrape_configs:
 		t.Fatalf("cannot parse data: %s", err)
 	}
 	swsNew := cfgNew.getFileSDScrapeWork(sws)
-	if equalStaticConfigForScrapeWorks(swsNew, sws) {
+	if equalScrapeWorks(swsNew, sws) {
```
Comment (xbsura) [inline lib/promscrape/config_test.go:100]: make ScrapeWork equal check more strict

File: lib/promscrape/config.go [hunk 3]
```diff
@@ -143,30 +143,6 @@ func unmarshalMaybeStrict(data []byte, dst interface{}) error {
 	return err
 }
 
-func (cfg *Config) kubernetesSDConfigsCount() int {
```
Comment (xbsura) [inline lib/promscrape/config.go:146]: useless, delete it

File: lib/promscrape/scraper.go [hunk 4]
```diff
@@ -78,337 +81,347 @@ func runScraper(configFile string, pushData func(wr *prompbmarshal.WriteRequest)
...
+		EC2SDReloadCh := make(chan struct{})
+		GCESDReloadCh := make(chan struct{})
+
 		stopCh := make(chan struct{})
 		var wg sync.WaitGroup
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
-			runStaticScrapers(cfg, pushData, stopCh)
+			runSDScrapers(Static, cfg, pushData, staticReloadCh, stopCh)
```
Comment (xbsura) [inline lib/promscrape/scraper.go:95]: all SD Scraper run in one function

File: lib/promscrape/scraper.go [hunk 5]
```diff
@@ -78,337 +81,347 @@ func runScraper(configFile string, pushData func(wr *prompbmarshal.WriteRequest)
 
 	mustStop := false
 	for !mustStop {
+		staticReloadCh := make(chan struct{})
+		fileSDReloadCh := make(chan struct{})
+		kubernetesSDReloadCh := make(chan struct{})
+		EC2SDReloadCh := make(chan struct{})
+		GCESDReloadCh := make(chan struct{})
```
Comment (xbsura) [inline lib/promscrape/scraper.go:88]: separt restart and reload chanel

File: lib/promscrape/scraper.go [hunk 6]
```diff
@@ -78,337 +81,347 @@ func runScraper(configFile string, pushData func(wr *prompbmarshal.WriteRequest)
...
+			runSDScrapers(GCESD, cfg, pushData, GCESDReloadCh, stopCh)
 		}()
 
+		reloadConfig := func() {
+			cfgNew, _, err := loadConfig(configFile)
+			if err != nil {
+				logger.Errorf("cannot read %q on SIGHUP: %s; continuing with the previous config", configFile, err)
+				return
+			}
+			*cfg = *cfgNew
```
Comment (xbsura) [inline lib/promscrape/scraper.go:124]: use pointer instead value to pass update

File: lib/promscrape/scraper.go [hunk 7]
```diff
@@ -78,337 +81,347 @@ func runScraper(configFile string, pushData func(wr *prompbmarshal.WriteRequest)
...
+			staticReloadCh <- struct{}{}
+			fileSDReloadCh <- struct{}{}
+			kubernetesSDReloadCh <- struct{}{}
+			EC2SDReloadCh <- struct{}{}
+			GCESDReloadCh <- struct{}{}
```
Comment (xbsura) [inline lib/promscrape/scraper.go:129]: notify all sws reload config

**Comment by valyala** [comment]
@xbsura , I pushed the commit 3b234d82e541395628951db1b70f0d5079ae9918 , which is based on the ideas from the https://github.com/VictoriaMetrics/VictoriaMetrics/pull/449 . The commit covers this PR too. So closing it.

Thanks for you help and initial ideas!

**Pull request closed without merging**